feat(ci): adding renovatebot for dependency management

hjoshi123 opened the pull request:
<!--

Thanks for opening a pull request! Here are some tips to get everything merged smoothly:

1. Read our contributor guidelines: https://cert-manager.io/docs/contributing/

2. Make sure your commits are signed off: https://cert-manager.io/docs/contributing/sign-off/

3. If the PR is unfinished, raise it as a draft or prefix the title with "WIP:" so it's clear to everyone.

4. Be sure to allow edits from maintainers so it's easier for us to help: https://help.github.com/en/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork

-->

### Pull Request Motivation

<!-- Explain the motivation behind this PR. If there's a related issue or PR, link to it here! -->

As discussed in our biweekly meetings, we are planning to use selfhosted renovate in cert-manager repo currently and eventually in all of our cert manager projects. https://github.com/cert-manager/cert-manager/issues/7878

### Kind

<!--
The kind(s) listed after "kind" after this comment will be used by a bot to add labels when the PR is opened.
If omitted at PR creation, someone will need to make a new comment with them later (editing the description after the fact will not trigger the bot).
-->
/kind feature
<!--

Pick the kind(s) which best describe your PR from the following list:

	<cleanup | bug | feature | documentation | design | flake>

If you're unsure which is best or if you're not sure what we mean by "kind",
just ignore this section and a maintainer will fill it in for you!
-->

### Release Note

<!--

Should we mention this PR in release notes? If so, replace "NONE" with a line of text explaining what changed!

For more details, see: https://git.k8s.io/community/contributors/guide/release-notes.md

-->

```release-note
NONE
```

hjoshi123 commented:
/ok-to-test
/hold

hjoshi123 commented on renovate.json:
```diff
@@ -0,0 +1,161 @@
+      "matchPackagePrefixes": ["github.com/Azure{/,}**"]
+    },
+    {
+      "description": "Group Golang.rog/x packages",
```
oh yeah.. 😅

hjoshi123 commented on renovate.json:
```diff
@@ -0,0 +1,161 @@
+      "rangeStrategy": "pin"
+    },
+    {
+      "description": "Update Go version in go.mod",
```
Yup looks like the patch_go_version enables the upgrade of go versions.. will remove it from the scope of renovate

hjoshi123 commented on renovate.json:
```diff
@@ -0,0 +1,161 @@
+  "constraints": {
+    "go": ">=1.21"
+  }
+}
```
Oh that's an excellent point.. didnt think of it.. yes they do support .json5.. I can rewrite it; might need some time to test it again

hjoshi123 replied:
Done its now using json5

erikgb replied:
Thanks!

hjoshi123 commented:
@erikgb / @ThatsMrTalbot I wanted to also confirm that we have the test stuff as required status checks in github right?

erikgb commented:
> I wanted to also confirm that we have the test stuff as required status checks in github right?

Prow is running most of our CI in this project, and test jobs are required to merge + plus a lot of other checks/guardrails. Or were you thinking of some special situation?

hjoshi123 commented:
> > I wanted to also confirm that we have the test stuff as required status checks in github right?
> 
> 
> 
> Prow is running most of our CI in this project, and test jobs are required to merge + plus a lot of other checks/guardrails. Or were you thinking of some special situation?

Yes that works. Because if there aren't required checks then it merges the PR without tests. So wanted to make sure.

hjoshi123 commented on .github/workflows/renovate.yml:
```diff
@@ -0,0 +1,33 @@
+  workflow_dispatch:
+  push:
+  schedule:
+    - cron: "* * * * 0,6"
+    - cron: "* 22-23,0-4 * * 1-5"
```
Yes by default fork processing is disabled. I had to enable to run it on my fork but its generally disabled.

hjoshi123 replied:
The reasoning behind the cron jobs was that.. it runs on a reasonable time so that reviewers can review it the next day. But since we have auto merge on many of those maybe it makes sense it to remove it and run periodically every night.

hjoshi123 commented on .github/workflows/renovate.yml:
```diff
@@ -0,0 +1,33 @@
+        # see https://github.com/actions/checkout/issues/701 for extra info about this option
+        with: { fetch-depth: 0 }
+
+      - name: Self-hosted Renovate
```
That's what their "Definition" of self-hosted is... the official docs treats github actions as a self hosted option.

hjoshi123 commented on .github/workflows/renovate.yml:
```diff
@@ -0,0 +1,33 @@
+jobs:
+  renovate:
+    runs-on: ubuntu-latest
+    permissions:
+      contents: write
+      issues: write
+      pull-requests: write
+      id-token: write
```
Not sure about the id tokens (i can remove it) but the issue write is needed for dep dashboard. https://docs.renovatebot.com/configuration-options/#dependencydashboard . The app is not required as dependency dashboard is just an issue on github where it will track the open PRs which can be approved etc.

hjoshi123 commented on .github/workflows/renovate.yml:
```diff
@@ -0,0 +1,33 @@
+      - name: Self-hosted Renovate
+        uses: renovatebot/github-action@v43.0.5
+        with:
+          configurationFile: renovate.json5
```
yup that can be done

hjoshi123 commented:
@erikgb / @ThatsMrTalbot quick update on the PR so that things dont get lost

* Moved the config to .github
* Changed the cron timing to run every night
* Removed id-token:write permissions

erikgb commented on .github/workflows/renovate.yml:
```diff
@@ -0,0 +1,30 @@
+        with: { fetch-depth: 0 }
+
+      - name: Self-hosted Renovate
+        uses: renovatebot/github-action@v43.0.5
```
```suggestion
        uses: renovatebot/github-action@85b17ebd5abf43d1c34c01bd4c8dbb8d45bbc2c7 # v43.0.7
```

We should always use pinned GHA to reduce the possibility of a security incident.

hjoshi123 replied:
Yup.. forgot about that

erikgb reviewed (commented):
/approve

Soft LGTM from me. We will probably need to adjust the configuration when we see how this works, but I am tempted to try this soon.

erikgb commented:
/unhold
